src/components/Tables/Table.tsx

```diff
@@ -0,0 +1,149 @@
+"use client";
+
+import React, { useState, useEffect } from "react";
+import DataTable, { TableColumn } from "react-data-table-component";
+import TableFilterBar, { FilterText } from "@/components/Tables/TableFilterBar";
+
+interface Row {
+    checkBoxKey: number;
+    [key: string]: string | number;
+}
+interface Headers {
+    [key: string]: string;
+}
+
+export interface Datum {
+    [key: string]: string | number | null;
+}
+
+interface Props {
+    data: Datum[];
+    headers: Headers;
+}
+
+const itemIncludesFilterText = (headers: Headers, item: Datum, filterText: FilterText): boolean => {
+    for (const key of Object.keys(headers)) {
+        if (
+            !(item[key] ?? "")
+                .toString()
+                .toLowerCase()
+                .includes((filterText[key] ?? "").toLowerCase())
+        ) {
+            return false;
+        }
+    }
+    return true;
+};
+
+const filteredNoNullItems = (headers: Headers, data: Datum[], filterText: FilterText): Row[] => {
+    let row = 0;
+    return data.reduce((filteredNoNullList: Row[], item: Datum, currentIndex: number) => {
+        if (itemIncludesFilterText(headers, item, filterText)) {
+            const noNullItem: Row = { checkBoxKey: currentIndex };
+            for (const key of Object.keys(headers)) {
+                noNullItem[key] = item[key] ?? "";
+            }
+            noNullItem.id = row;
+            filteredNoNullList.push(noNullItem);
+            row++;
+        }
+        return filteredNoNullList;
+    }, []);
+};
+
+const Table: React.FC<Props> = (props) => {
+    const [filterText, setFilterText] = useState<FilterText>({});
+
+    const [selectCheckBox, setSelectCheckBox] = useState(
+        new Array<boolean>(props.data.length).fill(false)
+    );
+    const [selectAllCheckBox, setSelectAllCheckBox] = useState(false);
+
+    const toggleAllCheckBox = (): void => {
+        setSelectCheckBox(new Array<boolean>(props.data.length).fill(!selectAllCheckBox));
+        setSelectAllCheckBox(!selectAllCheckBox);
+    };
+    const toggleOwnCheckBox = (checkBoxKey: number): void => {
+        const selectCheckBoxCopy = [...selectCheckBox];
+        selectCheckBoxCopy[checkBoxKey] = !selectCheckBoxCopy[checkBoxKey];
+        setSelectCheckBox(selectCheckBoxCopy);
+    };
+
+    // Fixes hydration error by re-rendering when the DOM is ready - https://nextjs.org/docs/messages/react-hydration-error
```

Comment:
is this still needed if we have another `useEffect` in use?

Reply:
Yes it is needed (just checked by removing it) - the issue isn't about whether there is useEffect in the function or not, the issue is the table should not be returned until the dom is loaded, which still has to be checked by an initial useEffect (as useEffect can only be run after the dom is loaded)